Enhance Sandworm behaviour

penev92 opened the pull request:
I took some time before filing the PR to split the work into (mostly) self-contained work items, each in its own commit. The commit messages should reflect what happens where.

As a side effect, worms should have no more problems with targets moving out of range during the attack, reaching solid ground or being picked up by a carryall.

P.S.: I intentionally didn't make any of the discussed changes for making worms "less aggressive", because these are behavior/logic changes, and those will be balancing changes. I will file a follow-up PR for them once this is done.

penev92 commented:
The main thing here is that worms no longer get stuck on a single target, but update their target constantly depending on how noisy and how far it is. The values there probably need tweaking.

obrakmann commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -30,6 +29,8 @@ public abstract class AttackBaseInfo : ITraitInfo
 		[Desc("Does the attack type require the attacker to enter the target's cell?")]
 		public readonly bool AttackRequiresEnteringCell = false;
 
+		public readonly bool IgnoresVisibility = false;
```
Please add a description.

penev92 replied:
Done.

obrakmann commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,105 @@
+			Game.Debug("(Sandworm {0}) is attacking actor ({1} {2}).", self.ActorID, desiredPair.Actor.Info.Name, desiredPair.Actor.ActorID);
+			var dist = (self.Location - desiredPair.Actor.Location).Length;
+			var str = desiredPair.Trait.CurrentNoise - desiredPair.Trait.NoiseAttenuation * dist;
+			Game.Debug("Target Noise level is {0}. Distance is {1}, noise strength is {2}.", desiredPair.Trait.CurrentNoise, dist, str);
```
Those two most certainly will have to go before merging. Just pointing it out so it doesn't get forgotten.

penev92 replied:
Definitely. They are there because someone (I think it was @pchote) told me to drop some debug messages so the reviewers can see what's going on.

pchote replied:
It wasn't me, but I approve of the idea.

obrakmann commented on OpenRA.Mods.D2k/Traits/WormFood.cs:
```diff
@@ -0,0 +1,53 @@
+		readonly WormFoodInfo info;
+
+		public int CurrentNoise { get; private set; }
+		public double NoiseAttenuation { get { return WormFoodInfo.Attenuation; } }
```
Why not `return info.Attenuation`?

penev92 replied:
Because it is a static field and you can't access it like that.

pchote commented on OpenRA.Mods.D2k/Traits/WormFood.cs:
```diff
@@ -0,0 +1,53 @@
+		public static HashSet<string> DangerousTerrain = new HashSet<string> { "Sand", "Dune", "Spice", "SpiceBlobs" };
+
+		[Desc("How much does the sound subside with each unit of distance.")]
+		public static double Attenuation = 0.36;
```
Two problems here:
- Making these static means that *all* actors will share the same value, which will be determined arbitrarily by the order that your unit definitions are parsed.
- Floating point numbers can't safely be used for gameplay code. You'll need to rewrite this as an integer - I suggest copying how the warhead Falloff works.

pchote replied:
These info fields should also all be readonly.

pchote commented on OpenRA.Mods.D2k/Traits/WormFood.cs:
```diff
@@ -0,0 +1,53 @@
+		public object Create(ActorInitializer init) { return new WormFood(this); }
+	}
+
+	class WormFood : ITick
```
WormFood isn't a very professional name.  I suggest something like `AttractsWorms`.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -37,6 +38,7 @@ public abstract class AttackBase : IIssueOrder, IResolveOrder, IOrderVoice, ISyn
 	{
 		[Sync] public bool IsAttacking { get; internal set; }
 		public IEnumerable<Armament> Armaments { get { return getArmaments(); } }
+		public bool IgnoresVisibility { get { return Info.IgnoresVisibility; } }
```
Info is already public, so callers may as well access `Info.IgnoresVisibility` directly.

pchote commented on OpenRA.Mods.D2k/Traits/World/WormManager.cs:
```diff
@@ -22,16 +22,17 @@ namespace OpenRA.Mods.D2k.Traits
 
-		[Desc("Average time (seconds) between worm spawn")]
+		[Desc("Average time (in seconds) between worm spawn")]
 		public readonly int SpawnInterval = 120;
```
Could we change this to ticks instead?

ScottNZ commented on OpenRA.Mods.D2k/Traits/WormFood.cs:
```diff
@@ -0,0 +1,53 @@
+	class WormFoodInfo : ITraitInfo
+	{
+		[Desc("List of terrains on which making noise can get you killed by a sandworm.")]
+		public static HashSet<string> DangerousTerrain = new HashSet<string> { "Sand", "Dune", "Spice", "SpiceBlobs" };
```
This should also just be a string array.

penev92 replied:
The `.Contains()` below would really work better with a HashSet, though.

pchote replied:
This is inconsistent with our other traits, and I don't think or FieldLoader even understands HashSets.

obrakmann commented:
Oh boy, you sure weren't kidding when you said this would make worms more aggressive. Well, at least towards harvesters.  They target them almost exclusively now (together with devastators and the occasional siege tank), even if the harvester is half a world away and the worm is surrounded by a whole army of lower tier units. They just go and make a straight beeline for the harvs. In other words, you can't even try and protect your harvesters by sacrificing other units as you can with the current implementation. I'm really not sure if this will help game-play.

penev92 commented:
Yeah, the Noise and attenuation values will have to be tweaked.
Also 
> P.S.: I intentionally didn't make any of the discussed changes for making worms "less aggressive", because these are behavior/logic changes, and those will be balancing changes. I will file a follow-up PR for them once this is done.

But these two would have to be in this PR.

penev92 commented:
So here's what's in the last update (last commit):
 - Changed the Noise values of actors and the NoiseAttenuation value (see [here](http://pastebin.com/JUvdhDA5) for reference) for balancing and to switch all to integers
 - Renamed the `WormFood` trait to `MakesNoise` (sounds more generic than `AttractsWorms`)
 - Added a `ReducesNoise` trait to house the `NoiseAttenuation` field and attached it to the world actor
 - Removed the `IgnoresVisibility` property of `AttackBase`
 - Switched worm spawn interval from seconds to ticks
 - Added a debug message for when a worm is idle and just wandering to make it easier to track

pchote commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,51 @@
+		public MakesNoise(MakesNoiseInfo info)
+		{
+			this.info = info;
+			dangerousTerrain = new HashSet<string>(info.DangerousTerrain);
```
This is a premature optimization, and adds unnecessary complexity.  DangerousTerrain will never have more than a handful of elements, so the performance difference of Contains between the HashSet and a raw array will be small.

pchote commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,51 @@
+
+			var terrainName = self.World.Map.GetTerrainInfo(self.Location).Type;
+			
+			CurrentNoise = dangerousTerrain.Contains(terrainName) ? info.NoiseIntensity : 0;
```
This doesn't need to run every tick, but only if the current location has changed.  That makes the difference between an array and HashSet even more negligible.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,110 @@
+
+			targetCountdown = SandwormInfo.TargetRescanIntervalInTicks;
+
+			var food = self.World.ActorsWithTrait<MakesNoise>();
```
This will loop over the majority of actors on the map, which isn't a great approach.  It would be better to query `world.ActorsInRange` within an appropriate radius, and then filter that based on the presence of MakesNoise.

pchote commented on OpenRA.Mods.D2k/Traits/World/ReducesNoise.cs:
```diff
@@ -0,0 +1,33 @@
+	class ReducesNoiseInfo : ITraitInfo
+	{
+		[Desc("How much does the sound subside with each unit of distance (each cell).")]
+		public readonly int NoiseAttenuation = 40;
```
It seems unnecessary to extract this into its own trait.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,110 @@
+			// Basically any food would do
+			var maxStrength = int.MinValue;
+
+			foreach (var traitPair in food)
```
The way I would do this:

1. Move `NoiseAttenuation` into `MakesNoiseInfo` and drop the unnecessary `Noise` prefix from that and `NoiseIntensity`.
2. Remove the `DangerousTerrain` property - we don't have a use case for units generating noise on terrain types that the worms can't travel on.
3. Define a `MaxSearchRadius` property on `SandwormInfo` with a sensible value, probably 10 - 20 cells.  Worms shouldn't care about things that are happening on the other side of the map.
4. Loop over `world.FindActorsInCircle(MaxSearchRadius)`.
  - If the actor doesn't generate noise or is on unpathable terrain, then skip it.
  - If `wormLocation == actorLocation`, then attack the current cell and abort further evaluation.
  - If not, then accumulate the contribution of the actor in a "direction of most noise" vector: `noise += actorNoise * (wormPosition - actorPosition) / actorAttenuation;`
5. If the noise vector is non-zero then move in that direction, otherwise move in a random direction.

penev92 commented:
This should cover @pchote's latest comments.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,36 @@
+		[Desc("How much noise does this actor produce.")]
+		public readonly int Intensity = 0;
+
+		[Desc("How much does the sound subside with each unit of distance (each cell).")]
```
I see no reason to say both "each unit of distance" and clarify "each cell", instead of just saying "with each cell".

penev92 replied:
Point taken.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,132 @@
+		public const string IdleSequenceName = "idle";
+
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public const int TargetRescanIntervalInTicks = 32;
```
`readonly` not `const`.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,132 @@
+			move = Exts.Lazy(self.Trait<IMove>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+			positionable = Exts.Lazy(self.Trait<IPositionable>);
```
None of these are guaranteed as you don't `Requires<T>` any of them.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,132 @@
+			CheckForTargets(self);
+		}
+
+		private bool CheckForTargets(Actor self)
```
This needs a better name.
`CheckForTargets` is a verb, something like `AnyValidTargets` would be better.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/World/WormManager.cs:
```diff
@@ -22,16 +22,17 @@ namespace OpenRA.Mods.D2k.Traits
-		[Desc("Average time (seconds) between worm spawn")]
-		public readonly int SpawnInterval = 120;
+		[Desc("Average time (in ticks) between worm spawn")]
+		public readonly int SpawnInterval = 3000;   // Approximately 2 minutes
```
`3000 / 25` is exactly 120 so this isn't approximate.
Comments regarding the default value should go in the `Desc` so it will be visible on [the trait docs page](https://github.com/OpenRA/OpenRA/wiki/Traits).

penev92 replied:
Well you can't really say that 25 ticks == 1 second. That's an approximation, since ticks have nothing to do with time. They are work items.

Phrohdoh replied:
While that is technically true there is no need to add complexity where it doesn't need to exist.

I also just noticed the wording "average time" here while this is in fact a set timeframe, not an average.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/World/WormManager.cs:
```diff
@@ -22,16 +22,17 @@ namespace OpenRA.Mods.D2k.Traits
 
 		public readonly string WormSignNotification = "WormSign";
 
+		[Desc("The actor's name")]
```
Actor to spawn? Worm manager actor?
(more informative please)

Phrohdoh commented:
While you are in this logic could you fix `public WormManager(WormManagerInfo info, Actor self)` to be `public WormManager(Actor self, WormManagerInfo info)`, please?

penev92 commented:
Will deal with squashing the fixup/update commits at a later point.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+		public readonly int MaxSearchRadius = 32;
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LauchAttackRange = 3;
```
Type `Lauch` -> `Launch`.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+{
+	class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>, Requires<AttackBaseInfo>
+	{
+		public const string IdleSequenceName = "idle";
```
`readonly` not `const`

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+			AnyValidTargets(self);
+		}
+
+		private bool AnyValidTargets(Actor self)
```
No explicit private please.

obrakmann commented on OpenRA.Mods.D2k/Activities/SwallowActor.cs:
```diff
@@ -70,6 +69,7 @@ bool WormAttack(Actor worm)
 				return false;
 
 			stance = AttackState.EmergingAboveGround;
+		    sandworm.IsAttacking = true;
```
spaces -> tabs

obrakmann commented on OpenRA.Mods.D2k/Activities/SwallowActor.cs:
```diff
@@ -109,6 +109,8 @@ public override Activity Tick(Actor self)
 			// Wait for the worm to get back underground
 			if (stance == AttackState.ReturningUnderground)
 			{
+			    sandworm.IsAttacking = false;
```
spaces -> tabs

penev92 commented:
Rebased and updated as suggested.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,36 @@
+	[Desc("This actor makes noise.")]
+	public class MakesNoiseInfo : ITraitInfo
+	{
+		[Desc("How much noise does this actor produce.")]
```
Two options for this and the desc below:
* `s/./?`
* Change this into a statement

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+	class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>, Requires<AttackBaseInfo>
+	{
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public readonly int TargetRescanIntervalInTicks = 32;
```
I would rename this to `TargetRescanInterval` as the desc lets us (and modders) know it is measured in ticks.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+			if (AnyValidTargets(self))
+				return;
+
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
```
This will need to be removed before the final +1 and merger.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+			AnyValidTargets(self);
+		}
+
+		bool AnyValidTargets(Actor self)
```
The logic here really needs to be split up.
Either put it all into tick or make `AnyValidTargets` return a valid state that the rest of the logic depends on.

obrakmann commented on OpenRA.Mods.D2k/Activities/SwallowActor.cs:
```diff
@@ -109,6 +109,8 @@ public override Activity Tick(Actor self)
+				sandworm.IsAttacking = false;
+
 				// There is a 50-50 chance that the worm would just go away
 				if (self.World.SharedRandom.Next() % 2 == 0)
```
Could you make that value an attribute of the worm manager and set it to 80% per default?

Phrohdoh replied:
I'll also ask for this to be configurable on said Manager.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/World/WormManager.cs:
```diff
@@ -27,15 +27,16 @@ class WormManagerInfo : ITraitInfo
-		public readonly int SpawnInterval = 120;
-
-		public readonly string WormSignNotification = "WormSign";
+		[Desc("Average time (in ticks) between worm spawn.")]
```
This isn't an average time.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/World/WormManager.cs:
```diff
@@ -84,6 +84,11 @@ public void Tick(Actor self)
 			AnnounceWormSign(self, wormLocations);
 		}
 
+		Actor GetRandomSpawnPoint(Actor self)
```
Nothing is changing with this move so it is just noise.

penev92 replied:
Weird why I did that.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("This actor makes noise.")]
```
I think this should be clarified to state that this is an in-game (logic) system.

This could be misinterpreted as a way to play audio files.

penev92 replied:
Good point. Not too easy to explain though.

Phrohdoh replied:
>Noise is an in-game logic system (currently used by the D2k worms), not a way to play audio files.

I know it is rudimentary and not so nice, but it does explain it.
Any better suggestions are welcomed.

Unit158 replied:
The best way to describe it is probably to say "Does this actor attract worms?"

The idea of noise attracting worms is lore specific, and probably should not be used for any technical writing.

On 2015-01-08, at 1:53 PM, Taryn Hill <notifications@github.com> wrote:

> In OpenRA.Mods.D2k/Traits/MakesNoise.cs:
> 
> > @@ -0,0 +1,36 @@
> > +#region Copyright & License Information
> > +/*
> > +* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
> > +* This file is part of OpenRA, which is free software. It is made
> > +* available to you under the terms of the GNU General Public License
> > +* as published by the Free Software Foundation. For more information,
> > +* see COPYING.
> > +*/
> > +#endregion
> > +
> > +using OpenRA.Traits;
> > +
> > +namespace OpenRA.Mods.D2k.Traits
> > +{
> > +	[Desc("This actor makes noise.")]
> Noise is an in-game logic system (currently used by the D2k worms), not a way to play audio files.
> 
> I know it is rudimentary and not so nice, but it does explain it.
> Any better suggestions are welcomed.
> 
> \
> Reply to this email directly or view it on GitHub.
>

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+		public readonly int MaxSearchRadius = 32;
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LaunchAttackRange = 3;
```
`IgnoreNoiseAttackRange` is a bit better IMO as it clearly illustrates the purpose.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LaunchAttackRange = 3;
+
+		public readonly string IdleSequenceName = "idle";
```
I don't feel `Name` is needed here.
But this isn't of much importance.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+			: base(self, info)
+		{
+			this.info = info;
+			mobile = Exts.Lazy(self.Trait<Mobile>);
```
Again, these traits aren't guaranteed to exist on this actor (you are just hoping here).

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,130 @@
+		public readonly int TargetRescanIntervalInTicks = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly int MaxSearchRadius = 32;
```
Could just make this a WRange (and add "in cells" to the desc) and avoid the `WRange.FromCells` below.

pchote commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,36 @@
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("This actor makes noise, which causes them to be targeted by actors with the Sandworm trait.")]
+	public class MakesNoiseInfo : ITraitInfo
```
I think I mentioned this earlier, but the argument I saw on the IRC logs over this description reinforces it: if you need a description to explain the basic functionality of a trait, then you probably need to rethink the name.  `AttractsWorms` would be a good one for this.

pchote commented on OpenRA.Mods.D2k/Traits/MakesNoise.cs:
```diff
@@ -0,0 +1,36 @@
+		public readonly int Intensity = 0;
+
+		[Desc("How much does the sound subside with each cell.")]
+		public readonly int Attenuation = 40;
```
40 what?

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,139 @@
+		public readonly int TargetRescanInterval = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly WRange MaxSearchRadius = WRange.FromCells(32);
```
This seems way too big to me.  For reference, if you have a 1920 x 1080 px monitor and center your view on the worm, then it will be scanning out past the edges of your visible screen.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,139 @@
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != Info.IdleSequence)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
```
This doesn't look right.  You shouldn't need to reach into the inner workings of the animation to solve whatever problem you are working around here.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,139 @@
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+		}
+
+		public new void OnBecomingIdle(Actor self)
```
Using `new` is also a sign of something weird in our code base.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,139 @@
+			targetCountdown = Info.TargetRescanInterval;
+
+			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, Info.MaxSearchRadius);
+			var noiseDirection = new CVec();
```
Use `CVec.Zero` instead of `new CVec()`.  But don't use that here either - you should be using a `WVec` instead.

penev92 replied:
What would that give me?

pchote replied:
Let me invert that question - what does `CVec` give you.  `WPos`/`WVec` are the standard types for doing gameplay calculations, and you should only use `CPos`/`CVec` if your logic specifically need them.

Using the extremely low-resolution cell types throws away most of the resolution in the `noise.Info.Intensity * distance / noise.Info.Attenuation;` calculation (which uses integers), which I expect may make the worm behave strangely in some edge cases.

penev92 replied:
OK, point taken.

pchote commented on OpenRA.Mods.Common/Traits/Wanders.cs:
```diff
@@ -14,9 +14,9 @@
+	public abstract class WandersInfo : ITraitInfo
 	{
-		public readonly int WanderMoveRadius = 10;
+		public readonly int WanderMoveRadius = 1;
```
Note to reviewers: test that this doesn't regress visceroids and other creeps.

obrakmann replied:
In TS it's only set for the Baby Visceroid, not for the adult ones.

obrakmann commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,139 @@
+		{
+			targetCountdown = Info.TargetRescanInterval;
+
+			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, Info.MaxSearchRadius);
```
Because you select your target from within this set, the sandworms don't actually need the `RevealsShroud` trait anymore, so you can get a little perf back.

penev92 replied:
Done.

obrakmann commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -210,6 +210,7 @@
     <Compile Include="Traits\ExternalCapturable.cs" />
     <Compile Include="Traits\ExternalCapturableBar.cs" />
     <Compile Include="Traits\ExternalCaptures.cs" />
+    <Compile Include="Traits\IgnoresCloak.cs" />
```
I'm pretty sure you don't need this anymore either because of the way the target selection works now.

penev92 commented:
Removed the `RevealsShroud` trait from the worms as suggested.

obrakmann commented:
Gameplay-wise I'm quite happy with this now.

Things left to do from my perspective:
- [ ] Fix wander radius of adult visceroids in TS
- [ ] Remove `RevealsShroud` from sandworms
- [ ] Remove debugging statements

:+1: when those are all ticked off.

penev92 commented:
Once all requested changes are cleared and someone else confirms, I will remove the debug messages and squash that "fixup" commit.

penev92 commented:
Rebased so we can testplay with this.

penev92 commented:
Addressed the `WVec` request.
Also reduced the noise infantry produce because yesterday's game proved them to be too efficient at luring worms away from your harvesters.
Previous:
>300  - Grenadier, Sardaukar, Engineer, Bazooka, Medic, Thumper infantry (undeployed)
200  - Rifle, Saboteur

New:
>180  - Grenadier, Sardaukar, Engineer, Bazooka, Medic, Thumper infantry (undeployed)
200  - Rifle, Saboteur

Also made worms targetable, so if you have enough firepower you don't *have* to run. Don't overestimate your forces though!

Phrohdoh commented:
:+1: for targetable worms.

Should worms stay dead? Or possibly have a maximum number of worms to ever live/spawn per map?

obrakmann commented:
Now that worms are killable, you'll need to implement the `INotifyKilled` interface to decrease the worm count in the worm manager.

penev92 commented:
@obrakmann damn, I forgot about that one. Maybe just leave it for after a playtest?

Phrohdoh commented:
They need to implement that if killable, otherwise remove targetability.

penev92 commented:
That was my thinking. Anyway, I implemented it.

obrakmann commented:
OK, I'm happy with it. Still needs the debugging statements removed and commits squashed.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -30,6 +29,9 @@ public abstract class AttackBaseInfo : ITraitInfo
 		[Desc("Does the attack type require the attacker to enter the target's cell?")]
 		public readonly bool AttackRequiresEnteringCell = false;
 
+		[Desc("Does not care about shroud or fog. Enables the actor to launch an attack against a target even if he has no visibility of it.")]
```
s/he/the attacking actor.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/AttractsWorms.cs:
```diff
@@ -0,0 +1,36 @@
+		[Desc("How much noise this actor produces.")]
+		public readonly int Intensity = 0;
+
+		[Desc("How much does the sound subside with each cell (ammount of intensity units lost per cell).")]
```
This is a question that is incorrectly ended with a `.`. Change this to a `?` or make this a statement.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,147 @@
+		public readonly WRange MaxSearchRadius = WRange.FromCells(20);
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly WRange IgnoreDistractionsAttackRange = WRange.FromCells(3);
```
I'm not a fan of "distractions", why not just use the word "noise"?

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,147 @@
+		public readonly WRange IgnoreDistractionsAttackRange = WRange.FromCells(3);
+
+		[Desc("A number representing the chance this actor has of disappearing after it attacks (in %).")]
+		public readonly int ChanceToDisappear = 80;
```
We know that an integer is a number.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,147 @@
+		readonly Lazy<AttackBase> attackTrait;
+
+		public bool IsAttacking;
+		public bool IsMovingTowardTarget;
```
These should be set privately.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,147 @@
+				if (!actor.IsInWorld)
+					continue;
+
+				var noise = actor.TraitOrDefault<AttractsWorms>();
```
Wasn't AttractsWorms renamed to MakesNoise?

penev92 replied:
It was. And then it was reverted.

Phrohdoh replied:
:sad:

penev92 commented:
Addressed some requests. Also removed the debug messages because test plays.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,145 @@
+					return;
+				}
+
+				noiseDirection += noise.Info.Intensity * distance / noise.Info.Attenuation;
```
This doesn't look right at all.  The noise contribution *increases* with distance!

pchote replied:
You'll need to divide through by `distance.LengthSquared` to give a 1/r falloff, or `distance.Length * 3` for a more physically correct 1/r^2

pchote replied:
the Length calculation is non-trivial, so you would be best to calculate it once after distance, and then use that value.

ScottNZ commented on OpenRA.Mods.D2k/Traits/AttractsWorms.cs:
```diff
@@ -0,0 +1,36 @@
+		[Desc("How much noise this actor produces.")]
+		public readonly int Intensity = 0;
+
+		[Desc("How much the sound subsides with each cell (ammount of intensity units lost per cell).")]
```
`amount`

penev92 commented:
OK, so replaced the dynamic noise calculations via scary formulas with something easier to get, copying from SpreadDamageWarhead.

penev92 commented:
Squashed some commits.
I suppose we'll need @obrakmann's :+1: again after the recent noise changes.

Phrohdoh commented:
You'll have my :+1: once the commit messages are cleared up (discussed in IRC).
#7048 and WormSign when a worm is in under fog (#7170) will need to be fixed before a stable release.

penev92 commented:
Renamed a couple of things in `WormManager` to better reflect what they are/do.
Rebased so D2k can be properly tested after all the merging that was done after we tested yesterday.

Phrohdoh commented:
I've made some changes to this at https://github.com/Phrohdoh/OpenRA/tree/sandworms but the game crashes when the world starts so feel free to inspect it and take from it if you would like.

I think managing worms should be done this way instead of a raw count integer.

penev92 committed:
Add IgnoresCloak trait

penev92 committed:
Add IgnoresVisibility field to AttackBase

penev92 committed:
Enhance sandworm targeting logic by having actors produce noise

penev92 commented:
Tested this in multiplayer and I can say it feels good. Worms aren't a *major* issue, while still adding to the feel of the game.

penev92 commented:
By @pchote's request I changed the noise falloff values. I am still not 100% what he wanted, but I'm hopeful I got it right. :D

pchote commented on OpenRA.Mods.D2k/Traits/AttractsWorms.cs:
```diff
@@ -0,0 +1,60 @@
+		public readonly int[] Falloff = { 100, 100, 25, 11, 6, 4, 3, 2, 1, 0 };
+
+		[Desc("Range between falloff steps.")]
+		public readonly WRange RangeStep = new WRange(4096);
```
There is a mismatch between this, which defines sound out to ~36 cells from an actor, and `SandwormInfo.MaxSearchRadius` which cuts off the search radius at 20 cells.

penev92 replied:
Right. I saw that yesterday while testing, but for some reason couldn't change it while in debug. Later I just forgot :|

penev92 replied:
Set both to 27 cells so we don't get too greedy and choke on that `ActorsInRange`.

pchote commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -0,0 +1,159 @@
+		public override void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != Info.IdleSequence)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
```
You probably want to be using the more standard `PlayRepeating` here instead of `ReplaceAnim`.

pchote commented on mods/d2k/rules/arrakis.yaml:
```diff
@@ -35,9 +35,7 @@ SANDWORM:
-		TargetTypes: Underground
-	RevealsShroud:
-		Range: 32c0
+		TargetTypes: Ground
```
IIRC the worms in the original could only be attacked when they were above ground.  We can adjust this in a future pr.

penev92 replied:
I am not sure what you (and @Phrohdoh) mean by "above ground". Do you mean only when attacking? In the original you could attack worms if you see the trail that represents them. I like to think of it like they were as in Emperor: Battle for Dune when fremen are riding them - partially sticking out of the ground.

pchote replied:
I just tested this and you're right, so feel free to ignore my comment.

pchote commented:
I tested this in game, and it feels much better - good job!  It looks like you forgot to hook up the death animation (or maybe the worm timed out and despawned?), but that can be left to a future PR.

:+1: after fixing the minor issues noted above and rebase.

penev92 commented:
Rebased and updated.
I doubt we have a worm death animation. In the original the worm would just go underground after taking a certain amount of damage (was very nicely done using the engine-generated movement animation).
There is an urban legend that if you kill a worm **while it is attacking and is mid-air** it would leave a corpse like one of the 2 (or was it 3?) dead worm decoration tiles, but I've never found evidence to prove that.

obrakmann commented:
Is this a bug in the tileset definitions?

![screenshot from 2015-01-14 19 05 12](https://cloud.githubusercontent.com/assets/4331210/5744377/aca6b7ac-9c20-11e4-81a7-96cbc6124357.png)

obrakmann commented:
```
Exception of type `System.InvalidOperationException`: Attempted to get trait from destroyed object (bazooka 707 (not in world))
  at OpenRA.TraitDictionary.CheckDestroyed (OpenRA.Actor actor) [0x00026] in /home/oliver/devel/openra/git/OpenRA.Game/TraitDictionary.cs:85 
  at OpenRA.TraitDictionary.Contains[IgnoresCloak] (OpenRA.Actor actor) [0x00002] in /home/oliver/devel/openra/git/OpenRA.Game/TraitDictionary.cs:89 
  at OpenRA.Actor.HasTrait[IgnoresCloak] () [0x0000d] in /home/oliver/devel/openra/git/OpenRA.Game/Actor.cs:204 
  at OpenRA.Mods.Common.Traits.TargetableUnit.TargetableBy (OpenRA.Actor self, OpenRA.Actor viewer) [0x0000d] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Traits/TargetableUnit.cs:41 
  at OpenRA.Traits.Target.IsValidFor (OpenRA.Actor targeter) [0x00031] in /home/oliver/devel/openra/git/OpenRA.Game/Traits/Target.cs:86 
  at OpenRA.Mods.Common.Effects.Missile.Tick (OpenRA.World world) [0x0003d] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Effects/Missile.cs:134 
  at OpenRA.World.<Tick>m__4 (IEffect e) [0x00002] in /home/oliver/devel/openra/git/OpenRA.Game/World.cs:283 
  at OpenRA.WorldUtils.DoTimed[IEffect] (IEnumerable`1 e, System.Action`1 a, System.String text) [0x00022] in /home/oliver/devel/openra/git/OpenRA.Game/WorldUtils.cs:92 
```

obrakmann commented:
:+1: once the `IsDead` check in  is added in `TargetableBy`

penev92 commented:
Done.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/6d324b4a9407b441f474ba0a64600c1eaeca94bd%5E...6d324b4a9407b441f474ba0a64600c1eaeca94bd)

penev92 commented:
OK, so next on the agenda are #7170, which is an engine feature, but was inspired by the sandworms (and they need it to polish them) and #7130, which was waiting for this.
